src/libwalletqt/WalletManager.cpp

```diff
@@ -539,22 +534,15 @@ WalletManager::~WalletManager()
     m_scheduler.shutdownWaitForFinished();
 }
 
-void WalletManager::onWalletPassphraseNeeded(Monero::Wallet *)
+void WalletManager::onWalletPassphraseNeeded(bool on_device)
 {
-    m_mutex_pass.lock();
-    m_passphrase_abort = false;
-    emit this->walletPassphraseNeeded();
-
-    m_cond_pass.wait(&m_mutex_pass);
-    m_mutex_pass.unlock();
+    emit this->walletPassphraseNeeded(on_device);
 }
 
-void WalletManager::onPassphraseEntered(const QString &passphrase, bool entry_abort)
+void WalletManager::onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort)
 {
-    m_mutex_pass.lock();
-    m_passphrase = passphrase;
-    m_passphrase_abort = entry_abort;
-
-    m_cond_pass.wakeAll();
-    m_mutex_pass.unlock();
+    if (m_passphraseReceiver != nullptr)
+    {
+        m_passphraseReceiver->onPassphraseEntered(passphrase, enter_on_device, entry_abort);
```

Comment:
Should protect `m_passphraseReceiver` with a mutex.

Reply:
That it does not become null after the check?
It is set only in `WalletManager::openWallet` and `WalletManager::createWalletFromDevice`, both are protected by locking `m_mutex`. 

The scenario this could happen is: GUI somehow triggers (without previous request) `onPassphraseEntered`, meanwhile one of the two functions ends, scope guard sets `m_passphraseReceiver` to null in race. 

Should we then to protect from this add a new mutex, that is called from the scope guard, before setting `m_passphraseReceiver` to null?

Reply:
> Should we then to protect from this add a new mutex, that is called from the scope guard, before setting `m_passphraseReceiver` to null?

Yes, please do

Reply:
ok done, pls check

Reply:
Is okay now